Backport docs for System.Diagnostics.Process

carlossanlop opened the pull request:
Backporting public API documentation from MS Docs into triple slash comments for the System.Diagnostics.Process assembly, which includes the System.Diagnostics and Microsoft.Win32.SafeHandles namespaces.

This namespace was interesting for this experiment because some types had cross-platform implementations. This meant the tool had to find all the possible compilations for the project (one for each platform) and make sure the exact same documentation was placed on top of all the different partial implementations.

It was also interesting to find so many remarks that contained elements that couldn't be converted from markdown to xml, so the tool had to detect those cases and keep the CDATA section unmodified.

The full description of the purpose of this PR is described in the issue [#44969 - Pilot a new process that extracts API documentation from source code](https://github.com/dotnet/runtime/issues/44969) under the section "Bring documentation from Docs to triple slash".

I ran the [porting tool](https://github.com/carlossanlop/DocsPortingTool/pull/26) using this command:

```
DocsPortingTool.exe \
-CsProj D:\runtime\src\libraries\System.Diagnostics.Process\src\System.Diagnostics.Process.csproj
-Docs D:\dotnet-api-docs\xml
-Direction ToTripleSlash
-IncludedAssemblies System.Diagnostics.Process
-IncludedNamespaces System.Diagnostics,Microsoft.Win32.SafeHandles
```

Please compare the ported docs with their source files in dotnet-api-docs:

- https://github.com/dotnet/dotnet-api-docs/tree/master/xml/System.Diagnostics
- https://github.com/dotnet/dotnet-api-docs/tree/master/xml/Microsoft.Win32.SafeHandles

carlossanlop committed:
Automated port results.

carlossanlop committed:
Turn on GenerateDocumentationFile.

carlossanlop committed:
SafeProcessHandle.ReleaseHandle add missing comments (not found in MS Docs).

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/Microsoft/Win32/SafeHandles/SafeProcessHandle.Unix.cs:
```diff
@@ -37,6 +37,9 @@ public sealed partial class SafeProcessHandle : SafeHandleZeroOrMinusOneIsInvali
 
         internal int ProcessId { get; }
 
+        /// <summary>Closes the handle.</summary>
+        /// <returns>On Windows, returns <see langword="true" /> if the handle is closed successfully; <see langword="false" /> otherwise. On Unix, always returns <see langword="true" /></returns>
+        /// <remarks>On Windows, to get extended error information, call <see cref="System.Runtime.InteropServices.Marshal.GetLastWin32Error()" />.</remarks>
```
I added this manually (and copied it to the Unix file). There were no docs for this API in MS Docs:  https://github.com/dotnet/dotnet-api-docs/blob/dc599c5db9e3464765b4e8c319bc22e525f4eebf/xml/Microsoft.Win32.SafeHandles/SafeProcessHandle.xml#L170-L172

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/Microsoft/Win32/SafeHandles/SafeProcessHandle.cs:
```diff
@@ -15,6 +15,7 @@
+    /// <summary>Provides a managed wrapper for a process handle.</summary>
     public sealed partial class SafeProcessHandle : SafeHandleZeroOrMinusOneIsInvalid
     {
         internal static readonly SafeProcessHandle InvalidHandle = new SafeProcessHandle();
```
Line 22: The public parameterless constructor does not show up in MS Docs, so there was no documentation to backport.
@joelmartinez do you know why this constructor can't be seen in dotnet-api-docs?
https://github.com/dotnet/dotnet-api-docs/blob/dc599c5db9e3464765b4e8c319bc22e525f4eebf/xml/Microsoft.Win32.SafeHandles/SafeProcessHandle.xml

carlossanlop replied:
This is a bug. I opened an issue for the Docs dev team to address this: https://dev.azure.com/ceapex/Engineering/_workitems/edit/374279  cc @TianqiZhang

carlossanlop committed:
SafeProcessHandle adjust remarks.

safern commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/DataReceivedEventArgs.cs:
```diff
@@ -14,6 +41,16 @@ internal DataReceivedEventArgs(string? data)
+        /// When the redirected stream is closed, a null line is sent to the event handler. Ensure your event handler checks the <see cref="System.Diagnostics.DataReceivedEventArgs.Data" /> property appropriately before accessing it. For example, you can use the static method <see cref="string.IsNullOrEmpty" /> to validate the <see cref="System.Diagnostics.DataReceivedEventArgs.Data" /> property  ...
+        /// ## Examples
+        /// The following code example illustrates a simple event handler associated with the <see cref="System.Diagnostics.Process.OutputDataReceived" /> event. The event handler receives text lines from the redirected <see cref="System.Diagnostics.Process.StandardOutput" /> stream, formats the text, and writes the text to the screen.
+        /// [!code-cpp[Process_AsyncStreams#4](~/samples/snippets/cpp/VS_Snippets_CLR/process_asyncstreams/CPP/datareceivedevent.cpp#4)]
```
Doesn't these have to be under a CDATA?

carlossanlop replied:
Yes, I'm working on this 😄 . The tool needs to detect code examples or includes or hyperlinks, and in those cases, preserve the markdown.

carlossanlop replied:
Fixed! The tool will avoid converting remarks to xmls if it finds examples, includes or hyperlinks.

carlossanlop committed:
DataReceivedEventArgs adjust remarks.

carlossanlop committed:
MonitoringDescriptionAttribute adjust remarks.

carlossanlop committed:
Process.NonUap adjust remarks and altmembers.

carlossanlop committed:
Process.Win32 adjust remarks.

carlossanlop committed:
Fix remarks when they have includes or examples.

carlossanlop committed:
Supress rule CA1200 and use O: prefix for method overload groups.

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,7 @@
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "https://github.com/dotnet/csharplang/issues/320", Scope = "type", Target = "~T:System.Diagnostics.Process")]
+[assembly: SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "https://github.com/dotnet/csharplang/issues/320", Scope = "member", Target = "~T:System.Diagnostics.ProcessStartInfo")]
```
I was given the suggestion to use the `O:` prefix for method overload groups and globally suppress rule CA1200 (which prevents using prefixes). This will ensure that when we send the documentation back to MS Docs, we don't lose the method group hyperlinks.

safern replied:
Why not suppress it instead in `Directory.Build.props`? That would be cleaner. By doing this, we would need to include a similar file change in all assemblies when we port the documentation.

carlossanlop replied:
I didn't know. Thanks for the suggestion.

carlossanlop replied:
Had a conversation with Santi, we will try to disable this globally in `eng/CodeAnalysis.ruleset`. If that doesn't work, we will use `<NoWarn>` in the global `Directory.Build.props`.

gewarren replied:
Ruleset files are deprecated, BTW. @mavasani - should the dotnet/runtime repo try to replace the https://github.com/dotnet/runtime/blob/master/eng/CodeAnalysis.ruleset file? Although I see that dotnet/roslyn has a small .ruleset file too: https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/CodeAnalysisRules.ruleset.

carlossanlop replied:
@gewarren We actively use `CodeAnalysis.ruleset` in the runtime repo. If we should move to a different method, then we should probably track that in a separate issue. I am not sure if we have a specific need for a ruleset file. Thanks for bringing it up, though.

I opened a PR to move CA1200 from Warning to Info: https://github.com/dotnet/runtime/pull/48612

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -6,8 +6,63 @@
 namespace System.Diagnostics
 {
+    /// <summary>Provides access to local and remote processes and enables you to start and stop local system processes.</summary>
+    /// <remarks><format type="text/markdown"><![CDATA[
```
@gewarren here's an example of a large remarks section that should be moved to an md file. I'll do that next.

carlossanlop replied:
I opened a PR in dotnet-api-docs, where I'm introducing the md files (since they will live in that repo). After we merge that, I'll re-run the tool to update these comments.
https://github.com/dotnet/dotnet-api-docs/pull/5360

jeffhandley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessThreadCollection.cs:
```diff
@@ -5,21 +5,16 @@
 namespace System.Diagnostics
 {
-    /// <devdoc>
-    ///    <para>[To be supplied.]</para>
```
`[To be supplied]` ✔️  😆

carlossanlop replied:
Those comments have been there since .NET Framework: https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/ProcessThreadCollection.cs,15

carlossanlop committed:
Remove GlobalSuppresions

carlossanlop committed:
Move lengthy remarks to md files in dotnet-api-docs.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ThreadState.cs:
```diff
@@ -3,49 +3,37 @@
-    /// </devdoc>
+    /// <summary>Specifies the current execution state of the thread.</summary>
+    /// <remarks><format type="text/markdown"><![CDATA[
+    /// > [!IMPORTANT]
```
I haven't kept up with discussion -- is this `<![CDATA[[!IMPORTANT]` style what we will use log term? Why not just `<important>` or something?

also by using CDATA, the internal tags are not validated (eg the `xref` tag here).

carlossanlop replied:
The consensus was that whenever the tool finds items that have no way of getting converted from markdown to xml, we will preserve the CDATA section, to avoid losing data.

For example: `[name](url)` or `[!INCLUDE[name](~/path/to/file)]` have no xml equivalent.

These are the only xml tags we support within a remarks section:
![image](https://user-images.githubusercontent.com/1175054/108793260-fa6c8880-7537-11eb-8cc1-930aeb3063dc.png)

You're correct, xrefs don't get validated in source, but consider that this is the backport of what we already had in MS Docs, so it's pre-validated. I expect devs to add future new remarks as xml comments.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Windows.cs:
```diff
@@ -6,23 +6,63 @@
+        /// <remarks>This property is referenced if the process is being started by using the user name, password, and domain.
+        /// If the value is <see langword="true" />, the user's profile in the `HKEY_USERS` registry key is loaded. Loading the profile can be time-consuming. Therefore, it is best to use this value only if you must access the information in the `HKEY_CURRENT_USER` registry key.
+        /// In Windows Server 2003 and Windows 2000, the profile is unloaded after the new process has been terminated, regardless of whether the process has created child processes.
+        /// In Windows XP, the profile is unloaded after the new process and all child processes it has created have been terminated.</remarks>
```
There is no particular need to clean up the docs while we do this, of course. But it is funny to see mention of things as old as Windows XP in here. I guess in general, we expect there to be comments like this in our code that do not apply to .NET Core at all?

(And what about comments on API that we simply don't have any more for some reason, that's got to be very rare of course.)

carlossanlop replied:
> we expect there to be comments like this in our code that do not apply to .NET Core at all?

Yes, I'm porting everything unmodified. But also keep in mind that MS Docs had one single docs section for APIs that exist in both .NET Core and .NET Framework. This particular API you pointed out exists in both versions, so it makes sense to preserve this comment.

For APIs that _only_ exist in .NET Framework, we decided their source of truth will be MS Docs, since we won't be able to backport it to source.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Win32.cs:
```diff
@@ -7,8 +7,36 @@
+        /// <remarks><format type="text/markdown"><![CDATA[
+        /// The <xref:System.Diagnostics.ProcessStartInfo.Verbs%2A> property enables you to determine the verbs that can be used with the file specified by the <xref:System.Diagnostics.ProcessStartInfo.FileName%2A> property. You can set the <xref:System.Diagnostics.ProcessStartInfo.Verb%2A> property to the value of any verb in the set. Examples of verbs are "Edit", "Open", "OpenAsReadOnly", "Prin ...
+        /// When you use the <xref:System.Diagnostics.ProcessStartInfo.Verbs%2A> property, you must include the file name extension when you set the value of the <xref:System.Diagnostics.ProcessStartInfo.FileName%2A> property. The file name extension determines the set of possible verbs.
+        /// ## Examples
```
Same question as below -- is there/will there be any kind of tooling validatoin of the markdown in these CDATA blocks?

carlossanlop replied:
Since it's being backported without changes from MS Docs, it has already been validated there.

We do not have plans to have markdown sections validated on the source side. Do you have strong opinions on getting validation? This would require opening a bug in the csharplang repo to request this feature, right?

danmoseley replied:
No I don't have a strong opinion. I doubt the compiler is ever going to validate MD/XML mixtures. I was more wondering whether we are setting ourselves up for success by mixing two markdown languages together and whether there is a path out of that. Clearly the api docs repo chose to do that but we have the luxury of time to be sure we have the right model.

carlossanlop commented:
@gewarren @joelmartinez do you know if mdoc knows how to interpret an `<example>` element? Can it be properly converted to something meaningful in the MS Docs xmls?

If it can be handled correctly, then in triple slash comments, we could separate remarks from examples, which would allow us to save the remarks as pure xml in more cases, while the examples would be saved as markdown:

#### We do this now
```cs
/// <remarks><format type="text/markdown"><![CDATA[
///
/// These are the remarks with a <xref:System.IO.File>.
///
/// ## Example
///
/// ```cs
/// File = new File();
/// ```
/// 
/// ]]></format></remarks>
```
#### Do this instead
```cs
/// <remarks>These are the remarks with a <see cref="System.IO.File" />.</remarks>
/// <example><format type="text/markdown"><![CDATA[
/// ```cs
/// File = new File();
/// ```
/// ]]></format></example>
```

danmoseley commented:
> Do this instead

That makes more sense to me - remarks in regular XML, with CDATA for code fenced blocks. As it is, in the Compression PR for example, I see a mixture of `<xref:System.IO.IOException>` and `cref="System.IO.IOException">` .. we aren't going to get it right going forward, not just because it's a mixture but also because the compiler can't validate the xref part.

carlossanlop commented:
@danmoseley I'm tempted to just do the change in DocsPortingTool to split examples this way, and at the same time make the request for the Docs dev team to address this in mdoc when we start sending them intellisense xml files containing `<example>` sections. I doubt we can handle them today.

@gewarren @joelmartinez @TianqiZhang @mimisasouvanh FYI

gewarren commented:
> > Do this instead
> 
> That makes more sense to me - remarks in regular XML, with CDATA for code fenced blocks. As it is, in the Compression PR for example, I see a mixture of `<xref:System.IO.IOException>` and `cref="System.IO.IOException">` .. we aren't going to get it right going forward, not just because it's a mixture but also because the compiler can't validate the xref part.

The `<example>` tag does render an "Examples" heading. Here are a couple of "examples":

- https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.framework.xamltypes.datasource.label?view=msbuild-16#Microsoft_Build_Framework_XamlTypes_DataSource_Label
- https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.beginscope?view=dotnet-plat-ext-5.0#examples (this one obviously needs some code fencing - I will fix it)

I'm wondering if we even need the CDATA for the example though. Can we just use `<code>` tags?

carlossanlop commented:
Thanks for confirming, @gewarren ! I will try to use `<code>` as you're suggesting, instead of CDATA. When the Docs CI starts working again, I'll submit a test PR to confirm it renders fine.

carlossanlop commented:
@gewarren I think `<code>` won't work: We need to specify the programming language and that can only be done with markdown. Even if we link to an external file with INCLUDE, we also need markdown. But having an `<example>` section will definitely help reduce the necessity of using markdown for a remarks section.

danmoseley commented:
What would it take to use a language attribute on the code tag? With a default treated as c#. Would it need a compiler change? We have time to get this right.

TianqiZhang commented:
Docs .Net pipeline does support <code> with language attribute, sorry that I'm OOF today so unable to grad an example right now, if you guys find any issue using this feature please feel free to reach me and I'll take a look ASAP.

carlossanlop commented:
I found it: `<code class="lang-csharp">`

carlossanlop commented:
@TianqiZhang do you know if `<a href="url" />` can be properly converted to a hyperlink by Docs?

TianqiZhang commented:
@carlossanlop right now `<a href="url" />` requires text in it otherwise it won't render properly.

carlossanlop commented on eng/CodeAnalysis.ruleset:
```diff
@@ -56,7 +56,7 @@
     <Rule Id="CA1069" Action="None" />          <!-- Enums values should not be duplicated -->
     <Rule Id="CA1070" Action="Info" />          <!-- Do not declare event fields as virtual -->
-    <Rule Id="CA1200" Action="Warning" />       <!-- Avoid using cref tags with a prefix -->
+    <Rule Id="CA1200" Action="Info" />          <!-- Avoid using cref tags with a prefix -->
```
This has already [been merged](https://github.com/dotnet/runtime/pull/48612). I need to rebase this PR.

carlossanlop committed:
Update remarks to be mostly xml. Minimize the markdown sections.

carlossanlop commented:
@danmoseley [I updated the tool](https://github.com/carlossanlop/DocsPortingTool/pull/27) so that it ports remarks as xml, but only the specific sections that cannot be translated to xml, are wrapped with cdata. I updated this PR to reflect those changes.

carlossanlop committed:
Fix build failures due to malformed comments.

carlossanlop committed:
DataReceivedEventArgs lengthy remarks

carlossanlop committed:
DataReceivedEventArgs lengthy remarks and fix paths.

carlossanlop committed:
ProcessAfinity lengthy remarks

carlossanlop committed:
** to <b>

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs:
```diff
@@ -10,6 +10,16 @@
+    /// <altmember cref="System.Diagnostics.Process.CloseMainWindow"/>
+    /// <altmember cref="O:System.Diagnostics.Process.Kill"/>
+    /// <altmember cref="System.Diagnostics.ProcessThread"/>
+    /// <related type="ExternalDocumentation" href="https://code.msdn.microsoft.com/windowsdesktop/Using-the-NET-Process-Class-d70597ef">Using the .NET Process Class</related>
```
dead link

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -15,12 +15,16 @@
+    /// <altmember cref="System.Diagnostics.Process.CloseMainWindow"/>
+    /// <altmember cref="O:System.Diagnostics.Process.Kill"/>
+    /// <altmember cref="System.Diagnostics.ProcessThread"/>
+    /// <related type="ExternalDocumentation" href="https://code.msdn.microsoft.com/windowsdesktop/Using-the-NET-Process-Class-d70597ef">Using the .NET Process Class</related>
```
dead link

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs:
```diff
@@ -28,51 +41,119 @@ public sealed partial class ProcessStartInfo
+        /// // or if you prefer collection property initializer syntax:
+        /// var info = new System.Diagnostics.ProcessStartInfo("cmd.exe")
+        /// {
+        /// ArgumentList = {
+        /// "/c",
+        /// "dir",
+        /// @"C:\Program Files\dotnet"
+        /// }
```
does code samples provided in the comments support indentation?
```suggestion
        /// ArgumentList = {
        ///     "/c",
        ///     "dir",
        ///     @"C:\Program Files\dotnet"
        /// }
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessPriorityClass.cs:
```diff
@@ -3,52 +3,28 @@
-        ///     operating system. Use extreme care when using the high-priority class, because
-        ///     a high-priority class application can use nearly all available CPU time.
-        /// </devdoc>
+        /// <summary>Specifies that the process performs time-critical tasks that must be executed immediately, such as the <see langword="Task List" /> dialog, which must respond quickly when called by the user, regardless of the load on the operating system. The threads of the process preempt the threads of normal or idle priority class processes. <br />Use extreme care when specifying <see lan ...
```
it's one of the places where having a word-wrap would be really nice

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessThread.Windows.cs:
```diff
@@ -94,12 +124,28 @@ private ThreadPriorityLevel PriorityLevelCore
+        /// |0x0004|3|
+        /// |0x0005|1 or 3|
+        /// |0x0007|1, 2, or 3|
+        /// |0x000F|1, 2, 3, or 4|
```
just for my education: is VS smart enough to format this table properly?

adamsitnik reviewed (approved):
This was... a LOT of changes! I am really happy to see the old comments gone! Great work @carlossanlop   👍 

I did not find any blocking issues, however, it would be nice if you could:
* fix|remove the dead links
* consider implementing word-wrap for the tool
* consider using a bot account for this kind of PRs (you are de facto becoming an owner of 20% of the code? FWIW I am fine with that, but in the past, we have used bots for similar things)

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/DataReceivedEventArgs.cs:
```diff
@@ -3,8 +3,26 @@
+    /// <param name="sender">The source of the event.</param>
+    /// <param name="e">A <see cref="System.Diagnostics.DataReceivedEventArgs" /> that contains the event data.</param>
+    /// <format type="text/markdown"><![CDATA[
+    /// [!INCLUDE[remarks] (~/includes/remarks/System.Diagnostics/DataReceivedEventHandler/DataReceivedEventHandler.md)]
```
this was answered offline, marking as resolved

gewarren commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -6,8 +6,34 @@
+        /// -or-
+        /// There is no process associated with this <see cref="System.Diagnostics.Process" /> object.
+        /// -or-
+        /// The calling process is a member of the associated process' descendant tree.</exception>
```
```suggestion
        /// The calling process is a member of the associated process's descendant tree.</exception>
```

gewarren commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -6,8 +6,34 @@
+        /// There is no process associated with this <see cref="System.Diagnostics.Process" /> object.
+        /// -or-
+        /// The calling process is a member of the associated process' descendant tree.</exception>
+        /// <exception cref="System.AggregateException">Not all processes in the associated process' descendant tree could be terminated.</exception>
```
```suggestion
        /// <exception cref="System.AggregateException">Not all processes in the associated process's descendant tree could be terminated.</exception>
```

gewarren commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs:
```diff
@@ -263,6 +283,22 @@ private bool CloseMainWindowCore()
+        /// If you have just started a process and want to use its main window title, consider using the <see cref="O:System.Diagnostics.Process.WaitForInputIdle" /> method to allow the process to finish starting, ensuring that the main window handle has been created. Otherwise, the system throws an exception.
+        /// <format type="text/markdown"><![CDATA[
+        /// > [!NOTE]
+        /// >  The main window is the window that currently has the focus; note that this might not be the primary window for the process. You must use the <xref:System.Diagnostics.Process.Refresh%2A> method to refresh the <xref:System.Diagnostics.Process> object to get the most up to date main window handle if it has changed.
```
```suggestion
        /// > The main window is the window that currently has the focus. This might not be the primary window for the process. Use the <xref:System.Diagnostics.Process.Refresh%2A> method to refresh the <xref:System.Diagnostics.Process> object to get the most up-to-date main window handle if it has changed.
```

gewarren commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessModule.cs:
```diff
@@ -5,52 +5,103 @@
+        /// <remarks><format type="text/markdown"><![CDATA[
+        /// The module's entry point is the location of the function that is called during process startup, thread startup, process shutdown, and thread shutdown. While the entry point is not the address of the DllMain function, it should be close enough for most purposes.
+        /// > [!NOTE]
+        /// >  Due to changes in the way that Windows loads assemblies, <xref:System.Diagnostics.ProcessModule.EntryPointAddress%2A> will always return 0 on [!INCLUDE[win8](~/includes/win8-md.md)] or [!INCLUDE[win81](~/includes/win81-md.md)] and should not be relied on for those platforms.
```
The INCLUDE files for Windows versions should be replaced.

gewarren commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessPriorityClass.cs:
```diff
@@ -3,52 +3,28 @@
-        ///     a real-time process that executes for more than a very brief interval can cause
-        ///     disk caches not to flush or cause the mouse to be unresponsive.
-        /// </devdoc>
+        /// <summary>Specifies that the process has the highest possible priority. <br />The threads of a process with <see langword="RealTime" /> priority preempt the threads of all other processes, including operating system processes performing important tasks. Thus, a <see langword="RealTime" /> priority process that executes for more than a very brief interval can cause disk caches not to fl ...
```
Should `<br />` be replaced with an actual newline?

gewarren commented:
> I found it: `<code class="lang-csharp">`

@TianqiZhang This doesn't actually seem to be the correct attribute - at least it doesn't give syntax highlighting [in this example](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.beginscope?view=dotnet-plat-ext-5.0#examples). According to https://ceapex.visualstudio.com/Engineering/_workitems/edit/227459/?triage=true, it seems the proper way to specify a language is like this:

```
/// <code language="csharp">
```
